tests/compliance/letexpr.json

```diff
@@ -0,0 +1,247 @@
+[
+ {
+  "given": {
+   "foo": {
+    "bar": "baz"
+   }
+  },
+  "cases": [
+   {
+    "expression": "let $foo = foo in $foo",
+    "result": {
+     "bar": "baz"
+    }
+   },
+   {
+    "expression": "let $foo = foo.bar in $foo",
+    "result": "baz"
+   },
+   {
+    "expression": "let $foo = foo.bar in [$foo, $foo]",
+    "result": [
+     "baz",
+     "baz"
+    ]
+   },
+   {
+    "command": "Multiple assignments",
+    "expression": "let $foo = 'foo', $bar = 'bar' in [$foo, $bar]",
+    "result": [
+     "foo",
+     "bar"
+    ]
+   }
+  ]
+ },
+ {
+  "given": {
+   "a": "topval",
+   "b": [
+    {
+     "a": "inner1"
+    },
+    {
+     "a": "inner2"
+    }
+   ]
+  },
+  "cases": [
+   {
+    "expression": "let $a = a in b[*].[a, $a, let $a = 'shadow' in $a]",
+    "result": [
+     [
+      "inner1",
+      "topval",
+      "shadow"
+     ],
+     [
+      "inner2",
+      "topval",
+      "shadow"
+     ]
+    ]
+   },
+   {
+    "comment": "Bindings only visible within expression clause",
+    "expression": "let $a = 'top-a' in let $a = 'in-a', $b = $a in $b",
+    "result": "top-a"
```

Comment:
This would enable mutually recursive definitions, analogous to letrec in other languages.  I'd rather have user defined functions be the mechanism to define that, if that ever ends up happening.

BTW, the best place for questions related to the spec would be on the [tracking proposal](https://github.com/jmespath/jmespath.jep/pull/18).

Reply:
If variables are set up in the same order they are declared would this be subject to recursive definition ?
I mean `$a` is assigned first, then `$b` comes after that and uses `$a`, etc...

Reply:
I see the comments on the tracking proposal linked, moving the discussion over there.